pallets/clusters/src/lib.rs

```diff
@@ -0,0 +1,479 @@
+			let current_block_number = <frame_system::Pallet<T>>::block_number().encode();
+			// Who is not concrete, so we need to encode it to get bytes
+			let who_bytes = who.encode();
+
+			let cluster_id = blake2_128(
+				&[&current_block_number, name.as_slice(), &extrinsic_index, &who_bytes].concat(),
+			);
+
+			// Check that the cluster does not exist already
+			ensure!(!<Clusters<T>>::contains_key(&cluster_id), Error::<T>::ClusterExists);
+
+			// At creation there are no roles and no members assigned to the cluster
+			let cluster = Cluster {
+				owner: who.clone(),
+				name: name.into_inner(), // `self.0`
+				cluster_id,
+				roles: vec![],
+				members: vec![],
+			};
+
+			let minimum_balance =
+				<pallet_balances::Pallet<T> as fungible::Inspect<T::AccountId>>::minimum_balance();
+			let origin_balance =
+				<pallet_balances::Pallet<T> as fungible::Inspect<T::AccountId>>::balance(&who);
+			ensure!(origin_balance > minimum_balance, Error::<T>::InsufficientFunds);
+
+			// write
+			// create an account for the cluster
+			let vault = get_vault_id(cluster_id);
+			<pallet_balances::Pallet<T> as fungible::Mutate<T::AccountId>>::mint_into(
+				&vault,
+				minimum_balance,
+			)?;
+
+			// create a pure proxy
+			pallet_proxy::Pallet::<T>::create_pure(
+				origin.clone(),
+				T::ProxyType::default(),
+				T::BlockNumber::zero(),
+				0,
+			)?;
+
+			<Clusters<T>>::insert(cluster_id, cluster);
+			// Don't clone, it's the last usage so let it move!
+			<ClustersByOwner<T>>::append(who, &cluster_id);
+
+			Self::deposit_event(Event::ClusterCreated { cluster_hash: cluster_id });
+			log::trace!("Cluster created: {:?}", hex::encode(&cluster_id));
+
+			Ok(())
+		}
+
+		/// Create a **Role** and assign it to an existing **Cluster**.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn create_role(
+			origin: OriginFor<T>,
+			cluster_id: Hash128,
+			role_name: BoundedVec<u8, T::NameLimit>,
+			settings: RoleSetting,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!cluster_id.len().is_zero(), Error::<T>::InvalidInput);
+			ensure!(!role_name.len().is_zero(), Error::<T>::InvalidInput);
+			ensure!(!settings.name.len().is_zero(), Error::<T>::InvalidInput);
+			ensure!(!settings.data.len().is_zero(), Error::<T>::InvalidInput);
+
+			let role_hash = blake2_128(&[&cluster_id.as_slice(), role_name.as_slice()].concat());
+
+			// Check that the caller is the owner of the cluster
+			let cluster = <Clusters<T>>::get(&cluster_id).ok_or(Error::<T>::ClusterNotFound)?;
+			ensure!(who == cluster.owner, Error::<T>::NoPermission);
+
+			// At creation there are no Members and no Rules assigned to a Role
+			let new_role = Role { name: role_name.into_inner(), members: vec![], rules: None };
+
+			// Check that the role does not exists already in the cluster
+			if <Clusters<T>>::get(&cluster_id)
+				.iter()
+				.any(|cluster| cluster.roles.iter().any(|role| new_role.name.eq(&role.name)))
```

Comment:
I don't think the first iter any is correct, doesn't get return just a single cluster?

Reply:
Oh my, yes of course. I have to stop coding at night, this is so stupid..